designs/aws/aws-launch-templates-v2.md

```diff
@@ -0,0 +1,146 @@
+
+The rationale we’ve stuck to so far is
+
+1. We only support basic use cases that should apply to most users. Categorizing a use case as basic has been somewhat arbitrary, but we’ve used upvotes on open issues and PRs to get that signal.
+2. At the minimum, functionality that other AWS provisioners like ManagedNodegroups give you without having to use a launch template should be available through Karpenter as well. Most of the parameters we’ve recently added fit this tenet.
+
+### How do we evolve LT usage?
+
+We’ve considered a few different approaches for this so far.
+
+**Approach 1 - Continue as is. Inline fields as necessary and enforce complete launch templates otherwise.**
+
+With this approach, we’ll continue to add overrides in our providerSpec based on user appetite, accepting PRs that solve a basic / common use case and asking all others to specify an entire launch template.
+
+
+*Pros*
+* Fastest path forward - we let users define which parameters are most important to them and let them unblock themselves with a PR if it should be beneficial to others.
+* Consider this as an example - EC2 has only added new fields twice in the last two years. Once for nitro enclaves, and the other for network index cards. Both of these aren't what we call a common use case, and even so it's not a lot to maintain. Network cards don't even apply to Karpenter (we tell customers to let the CNI configure network interfaces).
+  * Once we overcome the initial hump of missing fields, we shouldn't have frequent new things that are needed for a base case.
+
+*Cons*
+* The Provider API starts to have too many configuration knobs and becomes difficult to document and use.
+* Difficult messaging for users - how do we define what is and isn't supported on the provisioner? Forcing a user to define an entire launch template as an alternative is a lot of friction.
+
+
+**Approach 2 - Accept partial launch templates and fill in the rest.**
+
+This approach is inspired by Managed Nodegroups  - we let a customer only specify a certain set of properties in their launch template and let Karpenter automatically figure out the rest. That way a customer needing to only update the sizes of their root volumes needs to specify a much smaller and easier to maintain launch template like -
+
+```
+aws ec2 describe-launch-template-versions --launch-template-id lt-123
```

Comment:
is this a valid launch template?  We'd need to copy this data into our own launch template which would be difficult if we're using `$LATEST` or `$DEFAULT` to track the LT since it could change while we're copying.

Reply:
It is a valid launch template - but yes we'd need to copy that data into our own. 

> if we're using $LATEST or $DEFAULT to track the LT since it could change while we're copying.

I think that semantic isn't too surprising. If the alias they've given us to use changes which version it's pointing to, we'd change which LT we copy from too. If we're very concerned about this, we shouldn't let folks provide a version alias to Karpenter at all.

Fwiw, the pattern I've seen most folks adopt in the past is to let their infra tool resolve the alias and have an actual versionNumber passed to ASG / EC2 etc. So if you were using CloudFormation you do a `Fn::GetAtt MyLaunchTemplate.LatestVersionNumber` or something like that.

Reply:
Right that makes sense, I'll add it to the cons list. The time lag between the 'source' launch template being updated and us copying the updates over to the launch template we manage is likely going to be unpredictable - although I'm hoping it's in the order of ~seconds. We update our instance cache every 10 seconds, so maybe a LT cache in < 60 seconds wouldn't be too unreasonable.